url1d5e9d990f_Telsis_jOCP_tgz-pJ8-com_telsis_jocp_sampleApp_BasicOCPAppJ8/src/com/telsis/jocp/OCPSystemManager.java

```diff
@@ -0,0 +1,1344 @@
+         * thread terminates.
+         */
+        @Override
+        public void run() {
+            myThread = Thread.currentThread();
+            log.debug("SystemThread: thread starting");
+            watchdog.start();
+
+            synchronized (this) {
+                state = recalcMode();
+
+                // Bring up the OCP Links
+                for (Link link : links) {
+                    link.linkManager.connect(watchdog);
+                }
+
+                while (true) {
+                    watchdog.pat();
+                    internalRun();
+
+                    if (shutdownThread) {
+                        break;
+                    }
+
+                    try {
+                        this.wait(1 * MILLISECONDS);
+                    } catch (InterruptedException e) { // CSIGNORE: EmptyBlock
+                    }
+                }
+
+                log.debug("SystemThread: shutting down");
+                // Shutdown the OCP links
+                for (Link link : links) {
+                    link.linkManager.disconnect();
+                }
+            }
+            watchdog.stop();
+            log.debug("SystemThread: thread exiting");
+        }
+
+        /**
+         * Shut down the thread. This method waits for the thread to terminate
+         * before returning. If the thread has already terminated or has not
+         * been started, then this method will return immediately.
+         */
+        public void shutdownThread() {
+            log.debug("SystemThread: requesting shutdown");
+
+            synchronized (this) {
+                shutdownThread = true;
+                this.notify();
+            }
+
+            if (myThread != null) {
+                while (myThread.isAlive()) {
+                    try {
+                        myThread.join();
+                    } catch (InterruptedException e) { // CSIGNORE: EmptyBlock
+                    }
+                }
+            }
+            myThread = null;
+        }
+
+        /**
+         * Main loop procedure for the thread.
+         * <p/>
+         * This method is called every polling interval and carries out the
+         * following actions:
+         * <ol>
+         * <li>Each gapped link is checked to see if the gapping duration has
+         * elapsed, and if so gapping is disabled on that link.</li>
+         * <li>The last activity timestamp is checked on each link, and any
+         * links that have been inactive for more than the configured duration
+         * are marked as suspect.</li>
+         * <li>The system state is recalculated.</li>
+         * <li>The preferred unit is recalculated.</li>
+         * <li>If the system state is MISSING_MASTER and the configured timeout
+         * has elapsed, then the unit with the highest ID is sent a
+         * {@link BecomeMaster} request and the timeout is reset.
+         * </ol>
+         */
+        private void internalRun() {
```

Comment:
Recommendation generated by Amazon CodeGuru Reviewer. Leave feedback on this recommendation by replying to the comment or by reacting to the comment using emoji.

The cyclomatic complexity of this method is 12. By comparison, 98% of the methods in the CodeGuru reference dataset have a lower cyclomatic complexity. This indicates the method has a high number of decisions and it can make the logic difficult to understand and test. We recommend that you simplify this method or break it into multiple methods.